### PR TITLE
Display User

### PR DESCRIPTION
**Pull request by tpreusse**
add `id` to display user to enable client cache updates

### CODE REVIEW

File: graphql/resolvers/Comment.js [hunk 1]
```diff
@@ -97,13 +106,22 @@ module.exports = {
     const testimonial = !anonymous && await pgdb.public.testimonials.findOne({userId: commenter.id})
     const profilePicture = testimonial && testimonial.image
 
+    // TODO maybe use a HMAC
+    const hash = crypto
+      .createHash('sha256')
+      .update(`${discussion.id}${commenter.id}${USER_ID_SALT}`)
+      .digest('hex')
+    const id = fakeUUID(hash)
```
Comment (tpreusse) [inline graphql/resolvers/Comment.js]: why is `fakeUUID` necessary? The way I understand GraphQL `ID` supports any string. If not necessary I'd skip it as it looses some bytes in the process – probably making collisions more likely.

File: graphql/resolvers/Comment.js [hunk 2]
```diff
@@ -97,13 +106,22 @@ module.exports = {
     const testimonial = !anonymous && await pgdb.public.testimonials.findOne({userId: commenter.id})
     const profilePicture = testimonial && testimonial.image
 
+    // TODO maybe use a HMAC
```
Comment (tpreusse) [inline graphql/resolvers/Comment.js]: agree let's use HMAC:
```
crypto.createHmac('sha256', DISPLAY_AUTHOR_SECRET).update(`${discussion.id}${commenter.id}`).digest('hex')
```

File: graphql/resolvers/Comment.js [hunk 3]
```diff
@@ -1,5 +1,14 @@
 const Roles = require('../../lib/Roles')
 const createUser = require('../../lib/factories/createUser')
+const crypto = require('crypto')
+const fakeUUID = require('../../lib/fakeUUID')
+
+const {
+  USER_ID_SALT
```
Comment (tpreusse) [inline graphql/resolvers/Comment.js]: rename to `DISPLAY_AUTHOR_SECRET` – will help us to remember to keep it secret

**Comment by patte** [comment]
The HMAC approach provides no forward secrecy. If the DISPLAY_AUTHOR_SECRET should ever be leaked, all anonymous DisplayUser.ids could be guessed. It would require a serious bruteforce approach but it would be doable. We could
- leave it like this for now. env leaking would be catastrophic anyway.
- make a User.secretId. That we never directly expose.
- use DISPLAY_USER_SECRET + session id as key
what do you think @tpreusse

**Comment by tpreusse** [comment]
@patte DISPLAY_USER_SECRET + session id of the viewer would be perfect. But anon users don't have a sid right?

**Comment by patte** [comment]
@tpreusse true that. but IMHO that would be a good reason to enable uninitialised sessions, I think it makes for a perfect ephemeral user-bound secret (in the sense of visible to the server only) ID. If you don't object I'll enable it.

**Comment by tpreusse** [comment]
I'm against it. Creates complexity regarding privacy and caching. Also if somebody is able to steal our `DISPLAY_AUTHOR_SECRET` they can also steal our DB access and simply get the data.

Let's skip adding additional salts for now.

**Comment by patte** [comment]
Bear in mind that for subscriptions to work we enabled sticky-sessions for republik-backend. For this to work everybody gets a heroku-session-affinity cookie.
But it's true If you can steal the ENVs you have access to the DB. So let's skip it.

**Pull request merged**